src/Microsoft.AspNetCore.SignalR.Core/SerializedHubMessage.cs

```diff
@@ -35,21 +36,55 @@ public SerializedHubMessage(HubMessage message)
 
         public ReadOnlyMemory<byte> GetSerializedMessage(IHubProtocol protocol)
         {
+            // Double-check locking!
             if (!TryGetCached(protocol.Name, out var serialized))
             {
-                if (Message == null)
+                lock (_lock)
                 {
-                    throw new InvalidOperationException(
-                        "This message was received from another server that did not have the requested protocol available.");
-                }
+                    if (!TryGetCached(protocol.Name, out serialized))
+                    {
+                        if (Message == null)
+                        {
+                            throw new InvalidOperationException(
+                                "This message was received from another server that did not have the requested protocol available.");
+                        }
 
-                serialized = protocol.GetMessageBytes(Message);
-                SetCache(protocol.Name, serialized);
+                        serialized = protocol.GetMessageBytes(Message);
+                        SetCache(protocol.Name, serialized);
+                    }
+                }
             }
 
             return serialized;
         }
 
+        internal IEnumerable<SerializedMessage> GetAllSerializations()
```

Comment:
Add a comment that this is for unit tests so @davidfowl doesn't freak about the yield.